### DISCUSSION TITLE
Boolean expression parser

### DISCUSSION THREAD

**Discussion by test3211234** [winnow-rs/winnow#560] [Q&A]
Hi, I'm new to parser combinators so this is all pretty tough for me. I'm confused how to create a boolean expression parser like VSCode "when clause contexts": https://code.visualstudio.com/api/references/when-clause-contexts

I've been looking at this as the most similar but I'm still confused about the order and what functions to include what operators in and stuff: https://github.com/winnow-rs/winnow/blob/main/examples/arithmetic/parser_ast.rs

Right now I want this:
- Boolean variable (checked for value, implement myself)
- && , AND
- || , OR
- Variable (checked for value, implement myself) == string (enclosed in double quotes) or number , EQUALS
- Variable (checked for value, implement myself) != string (enclosed in double quotes) or number , NOT EQUALS
- ! , NOT
- ( and ) , PARENTHESES

I've written a handwritten parser for this before using Logos as the lexer. I'm wondering if this will be faster. I'm guessing the best approach is to store it into an AST and evaluate the AST everytime I want to check if the expression is true or false? Thanks Ed Page for the library

**Comment by epage** [comment]
> I've been looking at this as the most similar but I'm still confused about the order and what functions to include what operators in and stuff: 

Have you looked at the [tutorial](https://docs.rs/winnow/latest/winnow/_tutorial/chapter_0/index.html)?  It helps get you in the mindset which is that you only think about things in terms of the next byte(s) and not the grammar as a whole.

>  https://github.com/winnow-rs/winnow/blob/main/examples/arithmetic/parser_ast.rs

This parser is a good starting place as it is doing something similar to yours.  The main differences are
- unary operators: these will just be a special case of unary operators
- optional parenthesis: so long as everything as the same precedence and associativity, this isn't too different than [parsing json](https://docs.rs/winnow/latest/winnow/_topic/json/index.html) in that you can checking for individual expressions to follow *and* parenthesis is like checking for values and objects/arrays.

>  I'm guessing the best approach is to store it into an AST and evaluate the AST everytime I want to check if the expression is true or false? 

If you parse once and evaluate many times, having it in an AST or byte code would be faster than always re-parsing.
